[bugfix] solve crash when using `inspect()` on the "pyparsing" package

DrBenton opened the pull request:
## Type of changes

- [x] Bug fix - fixes #2284 
- [ ] New feature
- [ ] Documentation / docstrings
- [ ] Tests
- [ ] Other

:warning: This PR introduces a (minor) breaking change, in order to fix a (minor) issued described with the screenshots below. :point_down: 
But we can of course choose to keep the existing behaviour as-is instead :slightly_smiling_face: 

## Checklist

- [x] I've run the latest [black](https://github.com/psf/black) with default args on new code.
- [ ] I've updated CHANGELOG.md and CONTRIBUTORS.md where appropriate. `(N/A)`
- [x] I've added tests for new code.
- [x] I accept that @willmcgugan may be pedantic in the code review.

## Description

As reported [there](https://github.com/Textualize/rich/issues/2284), running this command was making Rich crash:
```py
python3 -c "import pyparsing; from rich import inspect; inspect(pyparsing, methods=True)"
```

It took me a while to understand what was going wrong, as the code from which the exception pops is in `Text.divide(offsets)` which is quite a complex part :exploding_head: 
But in the end I identified the culprits :ninja: : special ASCII characters! (in docstrings)

We were removing them from the content of the textual content of the `Text` class, but when caching the length of this content we were using the length of the non-sanitised text - which causes issues in some cases, as the text we're splitting into several lines doesn't actually have the length we think it has.

### Should we make control characters contained in docstrings visible in the inspection?

I first solved the bug by just making sure the cached length of the text is the one of the sanitised version of it, but then I noticed that the result was looking a bit odd with docstrings that contains such special characters, such as the `WordStart` class of this pyparsing package.

 - When dumped with `inspect` we were ending up with _"To emulate the ```` behavior of regular expressions"_, with 4 backticks in a row without content:
![Screenshot from 2022-05-25 10-56-56](https://user-images.githubusercontent.com/722388/170235974-023b89dc-bcd3-489a-aed7-077da98f36d9.png)

 - That's why I opted for another strategy: replacing these control codes with their "readable" equivalents, so `\b` is displayed as "\b":
![Screenshot from 2022-05-25 10-58-53](https://user-images.githubusercontent.com/722388/170236339-3ddc4c9d-dfb6-47b7-8f19-f5b86dd54882.png)

However, although it's rather unlikely I appreciate that doing such a change in Rich's behaviour could break people's code if they were using some parsing over the result of `rich.inspect` for example.
It depends I guess on whether or not we consider the stripping of control codes in docstrings, when inspecting code, as a feature of a bug? :thinking: 
If it's feature we may rather choose to keep the existing behaviour for the moment, and potentially change it later on with a major version bump, or an opt-in flag somewhere in Rich? :slightly_smiling_face:

DrBenton commented on rich/_inspect.py:
```diff
@@ -161,11 +155,9 @@ def safe_getattr(attr_name: str) -> Tuple[Any, Any]:
-            _doc = getdoc(obj)
+            _doc = self._get_formatted_doc(obj)
             if _doc is not None:
-                if not self.help:
```
The process of calling `cleandoc()` on the doctring's content and getting only its first paragraph if `self.help` is not True is now factorised in a single method, rather than done twice in this class

DrBenton commented on rich/control.py:
```diff
@@ -1,19 +1,35 @@
     13,  # Carriage return
 ]
-_CONTROL_TRANSLATE = {_codepoint: None for _codepoint in STRIP_CONTROL_CODES}
+_CONTROL_STRIP_TRANSLATE: Final = {
```
We pass this dict as a default value of the `_translate_table` (hidden) argument of the `strip_control_codes()` function, but dicts are mutable so we could theoretically have weird side-effects happening - which is picked up by some code analysis tools such as Pycharm's default inspections.
Maybe we should ask MyPy/Pyright/etc. to make sure that we never accidentally mutate these values, with such annotations?

DrBenton commented on rich/text.py:
```diff
@@ -141,15 +140,16 @@ def __init__(
         self.tab_size = tab_size
         self._spans: List[Span] = spans or []
-        self._length: int = len(text)
+        self._length: int = len(sanitized_text)
```
that's the bugfix itself! :slightly_smiling_face: i.e. we should use the length of the sanitised text, rather than the one of the text we received as an argument

DrBenton commented on rich/text.py:
```diff
@@ -394,9 +394,10 @@ def plain(self) -> str:
+            self._text[:] = [sanitized_text]
             old_length = self._length
-            self._length = len(new_text)
+            self._length = len(sanitized_text)
```
same bug here :slightly_smiling_face:

DrBenton commented on rich/text.py:
```diff
@@ -906,10 +907,10 @@ def append(
+                self._text.append(sanitized_text)
                 offset = len(self)
-                text_length = len(text)
+                text_length = len(sanitized_text)
```
...and here

willmcgugan commented on rich/control.py:
```diff
@@ -1,19 +1,35 @@
+    _codepoint: None for _codepoint in STRIP_CONTROL_CODES
+}
 
+_CONTROL_MAKE_READABLE_TRANSLATE: Final = {
```
Can we call this `CONTROL_ESCAPE`

willmcgugan commented on rich/control.py:
```diff
@@ -182,6 +198,22 @@ def strip_control_codes(
     return text.translate(_translate_table)
 
 
+def make_control_codes_readable(
```
I'd prefer `escape_control_code`. Would you mind changing the terminology from "make readable" elsewhere?

willmcgugan commented on rich/_inspect.py:
```diff
@@ -220,3 +210,12 @@ def safe_getattr(attr_name: str) -> Tuple[Any, Any]:
                 f"Run [b][magenta]inspect[/]([not b]inspect[/])[/b] for options."
             )
+
+    def _get_formatted_doc(self, object_: Any) -> Optional[str]:
```
Would you mind adding a docstring.

willmcgugan reviewed (changes requested):
Great. Just a few pendantic requests, and we should be good to merge.

DrBenton commented:
@willmcgugan "escape control codes" is a better terminology indeed! :slightly_smiling_face: 
Hopefully this commit addresses your comments: [25d0c0e](https://github.com/Textualize/rich/pull/2294/commits/25d0c0e89829685d2be17043812de82e97c460f7)

willmcgugan commented:
👍
